lib/transforms/addPrecacheServiceWorker.js

```diff
@@ -0,0 +1,69 @@
+var _ = require('lodash');
+var AssetGraph = require('../AssetGraph');
+var query = AssetGraph.query;
+var urlTools = require('urltools');
+var swPrecache = require('sw-precache');
+var Promise = require('bluebird');
+
+module.exports = function (queryObj) {
+    return function addPrecacheServiceWorker(assetGraph) {
+        return Promise.map(assetGraph.findAssets(_.extend({type: 'Html'}, queryObj)), function (htmlAsset) {
+            if (htmlAsset.isInline) {
+                throw new Error('addCacheManifestSinglePage: Cannot generate a CacheManifest for an inline Html asset');
+            }
+            var dynamicUrlToDependencies = {};
+            var assetByUrl = {};
+            assetGraph.eachAssetPostOrder(htmlAsset, {type: query.not(['HtmlAnchor', 'SvgAnchor', 'HtmlMetaRefresh', 'HtmlCacheManifest', 'HtmlConditionalComment', 'JavaScriptSourceUrl', 'JavaScriptSourceMappingUrl', 'JavaScriptSourceMap'])}, function (asset) {
+                // But only if the asset isn't inline, has been loaded, and isn't already in the manifest:
+                if (!asset.isInline && asset.isLoaded && asset !== htmlAsset) {
+                    var url = require('urltools').buildRootRelativeUrl(assetGraph.root, asset.url, assetGraph.root);
+                    assetByUrl[url] = asset;
+                    dynamicUrlToDependencies[url] = asset.isText ? asset.text : asset.rawSrc;
+                }
+            });
+            return Promise.fromNode(function (cb) {
+                swPrecache.generate({
+                    dynamicUrlToDependencies: dynamicUrlToDependencies,
+                    logger: function (message) {
+                        assetGraph.emit('info', htmlAsset.urlOrDescription + ': ' + message);
+                    }
+                }, cb);
+            }).then(function (serviceWorkerStr) {
+                serviceWorkerStr = serviceWorkerStr.replace(/^(var precacheConfig = )(\[[^\n]*\]);/m, function ($0, before, urlAndHashPairs) {
+                    return before + '[' + JSON.parse(urlAndHashPairs).map(function (urlAndHash) {
+                        return '[' + JSON.stringify(urlAndHash[0]) + '.toString(\'url\'),' + JSON.stringify(urlAndHash[1]) + ']';
+                    }).join(',') + '];';
+                });
+                var serviceWorkerUrl = htmlAsset.url.replace(/(?:\.\w+)?([?#]|$)/, '-service-worker.js');
+                var serviceWorker = new assetGraph.JavaScript({
+                    url: serviceWorkerUrl,
+                    text: serviceWorkerStr
+                });
+                serviceWorker.outgoingRelations.forEach(function (relation) {
+                    relation.to = assetByUrl[relation.href];
+                });
```

Comment:
Ah, turns out it's just to avoid running the `populate` transform because we already know exactly which assets will be referenced.

It's still ugly, though. I'll think about it.